oracle-plugin/docs/Oracle-batchsource.md

```diff
@@ -53,6 +53,37 @@ will be passed to the JDBC driver as connection arguments for JDBC drivers that
 **Default Row Prefetch:** The default number of rows to prefetch from the server.
 
 
+Data Types Mapping
+----------
+
+	| Oracle Data Type               | CDAP Schema Data Type        | Comment                |
+	| ------------------------------ | ---------------------------- | ---------------------- |
+	| VARCHAR2                       | Type.STRING                  |                        |
+	| NVARCHAR2                      | Type.STRING                  |                        |
+	| VARCHAR                        | Type.STRING                  |                        |
+	| NUMBER                         | LogicalType.DECIMAL          |                        |
+	| FLOAT                          | Type.DOUBLE                  |                        |
+	| LONG                           | Type.STRING                  |                        |
+	| DATE                           | LogicalType.TIMESTAMP_MICROS |                        |
+	| BINARY_FLOAT                   | Type.FLOAT                   |                        |
+	| BINARY_DOUBLE                  | Type.DOUBLE                  |                        |
+	| TIMESTAMP                      | LogicalType.TIMESTAMP_MICROS |                        |
+	| TIMESTAMP WITH TIME ZONE       | LogicalType.TIMESTAMP_MICROS | Converted to UTC time  |
+	| TIMESTAMP WITH LOCAL TIME ZONE | LogicalType.TIMESTAMP_MICROS |                        |
+	| INTERVAL YEAR TO MONTH         | Type.STRING                  |                        |
+	| INTERVAL DAY TO SECOND         | Type.STRING                  |                        |
+	| RAW                            | Type.BYTES                   |                        |
+	| LONG RAW                       | Type.BYTES                   |                        |
+	| ROWID                          | Type.STRING                  |                        |
+	| UROWID                         | Type.STRING                  |                        |
+	| CHAR                           | Type.STRING                  |                        |
+	| NCHAR                          | Type.STRING                  |                        |
+	| CLOB                           | Type.STRING                  |                        |
+	| NCLOB                          | Type.STRING                  |                        |
+	| BLOB                           | Type.BYTES                   |                        |
+	| BFILE                          | Type.BYTES                   | Java API is deprecated |
```

Comment:
An end user is not going to know what to do with the fact that the Java API is deprecated. Is BFile deprecated in Oracle  in general or just in the Java API?

Reply:
Comment deleted. Only `BFile` Java API is deprecated and I faced issues reading non-null values of this type. This data type contains a locator to a large binary file stored outside the database. There is no validation while creating these locators and users must ensure that the locator points to an existing file and database process has access to that file. 

I've managed to create such locators but faced issues reading them via JDBC connector. Should I investigate this deeper?

Reply:
What is a locator? Is it some custom class in the Oracle JDBC driver? A URI to a file that exists on the DB's local filesystem? A URI that points to an external file?

We definitely shouldn't be in a situation where the pipeline fails if it encounters a non-null value for a BFile column. If we can't get the actual content, then is there a string representation of the locator? What does that look like?

Reply:
BFILE is a data type used to store a locator (link) to an external file, which is stored outside of the database.
Values of this data type created as: 
```
CREATE TABLE files (ID INT, BFILE_COL BFILE);
CREATE DIRECTORY my_dir AS '/tmp';
INSERT INTO files VALUES (1, BFILENAME('MY_DIR', 'pic1.gif'));

```

We can select such values as:
```
SQL> select BFILE_COL from files;

BFILE_COL
--------------------------------------------------------------------------------
bfilename('MY_DIR', 'pic1.gif')
```

The [documentation contains](https://docs.oracle.com/cd/B19306_01/java.102/b14355/oralob.htm#CHDIGFBA) the following note: 
> You can create a BFILE. However, you cannot create an operating system file that a BFILE would refer to. Those are created only externally.

I finally managed to read BFILE locators via JDBC connector, so i added support of this data type for Source and Sink plugins(Yes, there is a custom [BFILE](https://docs.oracle.com/database/121/JAJDB/oracle/sql/BFILE.html) class in the Oracle JDBC driver, instances of which can be mapped to the `Type.BYTES`).

The current implementation allows to read and write locators(and not content of the external files), mapping BFILE to Type.BYTES.
Also, BFILE locator of non-existing file [added to Source/Sink tests](https://github.com/data-integrations/database-plugins/pull/46/files#diff-9c6424c7b677d0b14bfdab5253a3a18dR295).

Reply:
What is the byte value of a BFILE locator? That seems really odd to me. Is it https://docs.oracle.com/database/121/JAJDB/oracle/jdbc/OracleBfile.html#getBytes_long__int_ ? That seems like the actual file contents.

Reply:
I believe I used this method: https://docs.oracle.com/database/121/JAJDB/oracle/sql/Datum.html#getBytes__

Changed to use https://docs.oracle.com/database/121/JAJDB/oracle/jdbc/OracleBfile.html#getBinaryStream__ 
to retrieve the entire BFILE as a stream.